homework/05/ChatSys/src/Server.java

```diff
@@ -0,0 +1,135 @@
+    private ExecutorService pool;
+    private DateTimeFormatter formatter = DateTimeFormatter.ofPattern("HH:mm a", Locale.ENGLISH);
+
+    public Server(){
+        connections = new ArrayList<>();
+        done = false;
+    }
+
+    @Override
+    public void run(){
+        try {
+           server = new ServerSocket(4444);
+           pool = Executors.newCachedThreadPool();
+            while (!done) {                                 //assure there is constant connection
+                Socket client = server.accept();
+                ClientToServerConnection handler = new ClientToServerConnection(client);
+                connections.add(handler);
+                pool.execute(handler);
+            }
+        }catch (Exception e){
+            quit();
+        }
+    }
+
+    //to broadcast message to everyone
+    public void broadcast(String message){
+        for (ClientToServerConnection ch : connections){
+            if(ch!=null) {
+                ch.sendMessage(message);
+            }
+        }
+    }
+
+    public void quit() {                                            //shutdown of server
+        try {
+            done = true;
+            if (!server.isClosed()) {
+                server.close();
+            }
+            for (ClientToServerConnection ch : connections){
+                ch.quit();
+            }
+        }catch (IOException e){
+            //
+        }
+    }
+
+    class ClientToServerConnection implements Runnable{             //handles individual connections
+
+        private Socket client;
+        private BufferedReader in;                                  //get stream from the socket
+        private PrintWriter out;                                    // write to client
+        private String nickname;
+
+        public ClientToServerConnection(Socket client) {
+            this.client = client;
+        }
+
+        @Override
+        public void run(){
+            try {
+                    out = new PrintWriter(client.getOutputStream(), true);
+                    in = new BufferedReader(new InputStreamReader(client.getInputStream()));
+                    out.println("Enter nickname:");                 //message to the client
+                    nickname = in.readLine();                       //to be done with a command
+                    System.out.println(nickname + " connected");
+                    broadcast (nickname + " joined the chat!");      // to everyone
+
+                    String message;
+                    while ((message = in.readLine()) !=null){
+
+                        if (message.startsWith("/nick ")){          //to sign with nickname
+                            String[] messageSplit = message.split(" ", 2);
+                            if(messageSplit.length == 2){
+                                broadcast(nickname + "has new name" + messageSplit[1]);
+                                System.out.println(nickname + "has new name" + messageSplit[1]);

```

Comment:
Here you are copping code, which has a function in it. It is not an issue of grate importance, but you could save some processing power by making a string which contains the data. Then It can be is used by the functions in row 91 and 92